tests/unit/sharded_copying_manager_tests/copying_manager_test.py

```diff
@@ -0,0 +1,353 @@
+# Copyright 2014-2020 Scalyr Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from __future__ import unicode_literals
+from __future__ import absolute_import
+
+
+import time
+import shutil
+import os
+import platform
+import multiprocessing.managers
+import threading
+
+
+if False:
+    from typing import Dict
+    from typing import Tuple
+
+import pytest
+
+from scalyr_agent import scalyr_logging
+from scalyr_agent.configuration import Configuration
+
+from tests.unit.sharded_copying_manager_tests.common import (
+    CopyingManagerCommonTest,
+    TestableCopyingManager,
+    TestableCopyingManagerFlowController,
+    TestableCopyingManagerThreadedWorker,
+    TestableLogFile,
+    TestingConfiguration,
+)
+
+import mock
+
+import six
+
+log = scalyr_logging.getLogger(__name__)
+log.setLevel(scalyr_logging.DEBUG_LEVEL_0)
+
+
+def pytest_generate_tests(metafunc):
+    """
+    Run all tests for each configuration.
+    """
+    if "worker_type" in metafunc.fixturenames:
+        test_params = ["thread"]
+        if platform.system() != "Windows":
+            test_params.append("process")
+
+        metafunc.parametrize(
+            "worker_type, api_keys_count, workers_count",
+            [["thread", 1, 1], ["process", 1, 1], ["thread", 2, 2], ["process", 2, 2]],
+        )
+
+
+class CopyingManagerTest(CopyingManagerCommonTest):
+    @pytest.fixture(autouse=True)
+    def setup(self, worker_type, api_keys_count, workers_count):
+        super(CopyingManagerTest, self).setup()
+        self.use_multiprocessing_workers = worker_type == "process"
+        self.api_keys_count = api_keys_count
+        self.workers_count = workers_count
+
+    def teardown(self):
+        if self._instance is not None:
+            self._instance.stop_manager()
+
+            self._instance.cleanup()
+        super(CopyingManagerTest, self).teardown()
+
+    def _create_manager_instanse(
+        self,
+        log_files_number=1,
+        auto_start=True,
+        use_pipelining=False,
+        config_data=None,
+        disable_flow_control=False,
+    ):  # type: (int, bool, bool, Dict, bool) -> Tuple[Tuple[TestableLogFile, ...], TestableCopyingManager]
+
+        pipeline_threshold = 1.1
+        if use_pipelining:
+            pipeline_threshold = 0.0
+
+        initial_config_data = {
+            "debug_level": 5,
```

Comment:
Debug level 5 is fairly verbose.  Is that what you really want here (if it is, that's fine).  I just want to make sure it is not left over from some testing.